### PR TITLE
Features/ncne publish chart in caris

### PR DESCRIPTION
**Pull request by ssrajan**
When the Publish Chart button is clicked,
1. Validate the user and show the error message for invalid users
2. Show the publish chart modal dialog where the user can enter the chart version number and get additional details (show error message for invalid version numbers)
3. When publish button is pressed on the modal dialog,  publish the chart into Caris and display error message if any is returned.
4. If the Chart is published successfully,
    i) Complete the publish chart stage and stamp it with the username and the date
   ii) Add a system comment for the completion of stage.

### CODE REVIEW

File: src/Common.Helpers/CarisProjectHelper.cs [hunk 1]
```diff
@@ -18,6 +18,83 @@ public CarisProjectHelper(HpdDbContext hpdDbContext)
...
+
+            try
+            {
+
+                var connection = _hpdDbContext.Database.GetDbConnection();
+                await using var command = connection.CreateCommand();
+                command.CommandType = CommandType.Text;
+                command.CommandText = commandText;
+
+                connection.Open();
```
Comment (benbhall) [inline src/Common.Helpers/CarisProjectHelper.cs:38]:
```suggestion
                await connection.OpenAsync();
```

File: src/Common.Helpers/CarisProjectHelper.cs [hunk 2]
```diff
@@ -18,6 +18,83 @@ public CarisProjectHelper(HpdDbContext hpdDbContext)
...
+                await using var reader = command.ExecuteReader();
```
Comment (benbhall) [inline src/Common.Helpers/CarisProjectHelper.cs:39]:
```suggestion
                using var reader = await command.ExecuteReaderAsync();
```

File: src/Common.Helpers/CarisProjectHelper.cs [hunk 3]
```diff
@@ -18,6 +18,83 @@ public CarisProjectHelper(HpdDbContext hpdDbContext)
...
+
+                while (reader.Read())
```
Comment (benbhall) [inline src/Common.Helpers/CarisProjectHelper.cs:41]:
```suggestion
                while (await reader.ReadAsync())
```

File: src/Common.Helpers/CarisProjectHelper.cs [hunk 4]
```diff
@@ -18,6 +18,83 @@ public CarisProjectHelper(HpdDbContext hpdDbContext)
...
+
+        public async Task<bool> PublishCarisProject(int carisChartId)
+        {
+
+            var connection = _hpdDbContext.Database.GetDbConnection();
+            await using (var command = connection.CreateCommand())
+            {
+                try
+                {
+                    connection.Open();
```
Comment (benbhall) [inline src/Common.Helpers/CarisProjectHelper.cs:65]:
```suggestion
                    await connection.OpenAsync();
```

File: src/Common.Helpers/CarisProjectHelper.cs [hunk 5]
```diff
@@ -18,6 +18,83 @@ public CarisProjectHelper(HpdDbContext hpdDbContext)
...
+
+                    command.Transaction = connection.BeginTransaction();
+
+                    var commandText = "BEGIN P_SAVE_MANAGER.STARTNEWSAVE(null); " +
+                                      $"hpdowner.p_charts.publish ({carisChartId}); END;";
+
+                    command.CommandText = commandText;
+                    command.ExecuteNonQuery();
```
Comment (benbhall) [inline src/Common.Helpers/CarisProjectHelper.cs:73]:
```suggestion
                    await command.ExecuteNonQueryAsync();
```

File: src/Common.Helpers/CarisProjectHelper.cs [hunk 6]
```diff
@@ -18,6 +18,83 @@ public CarisProjectHelper(HpdDbContext hpdDbContext)
...
+
+                    command.Transaction.Commit();
```
Comment (benbhall) [inline src/Common.Helpers/CarisProjectHelper.cs:75]:
```suggestion
                    await command.Transaction.CommitAsync();
```

File: src/Common.Helpers/CarisProjectHelper.cs [hunk 7]
```diff
@@ -18,6 +18,83 @@ public CarisProjectHelper(HpdDbContext hpdDbContext)
...
+
+                }
+
+                catch (OracleException e)
+                {
+                    command.Transaction.Rollback();
```
Comment (benbhall) [inline src/Common.Helpers/CarisProjectHelper.cs:81]:
```suggestion
                   await command.Transaction.RollbackAsync();
```

File: src/Common.Helpers/CarisProjectHelper.cs [hunk 8]
```diff
@@ -18,6 +18,83 @@ public CarisProjectHelper(HpdDbContext hpdDbContext)
...
+                    var error = FormatOracleError(e);
+                    throw error;
+                }
+                catch (Exception)
+                {
+                    command.Transaction.Rollback();
```
Comment (benbhall) [inline src/Common.Helpers/CarisProjectHelper.cs:87]:
```suggestion
                  await command.Transaction.RollbackAsync();
```

File: src/Common.Helpers/CarisProjectHelper.cs [hunk 9]
```diff
@@ -18,6 +18,83 @@ public CarisProjectHelper(HpdDbContext hpdDbContext)
...
+        {
+
+            var connection = _hpdDbContext.Database.GetDbConnection();
+            await using (var command = connection.CreateCommand())
+            {
+                try
+                {
+                    connection.Open();
+
+                    command.Transaction = connection.BeginTransaction();
```
Comment (benbhall) [inline src/Common.Helpers/CarisProjectHelper.cs:67]:
```suggestion
                    command.Transaction = await connection.BeginTransactionAsync();
```

Comment (ssrajan) [reply src/Common.Helpers/CarisProjectHelper.cs]: Done

**Pull request merged**